doc/BluetoothLEStack.md

```diff
@@ -0,0 +1,171 @@
+
+### Advertising
+
+For advertising, the sytsem call interface allows a process to configure an
+advertising payload, advertising event type, scan response payload, interval and
+tx power. Permissible advertising types include:
+
+  * Connectable undirected
+
+  * Connectable directed
+
+  * Non-connectable undirected
+
+  * Scannable undirected
+
+The driver is _not_ responsible for validating that the payload for these
+advertising types follows any particular specification. Advertising event types
+that require particular interactions at the link-layer with peer devices (e.g.
+scanning or establishing connections) are not permissible:
+
+  * Scan request
+
+  * Scan response
+
+  * Connect request
+
+Scan response are sent automatically if a scan response payload is configured.
+Scan rerequest and connection requests are handled by other parts of the sytsem
+call interface.
+
+To set up an advertisement:
+
+  1. Configure the advertisement payload, type, interval, tx power and,
+     optionally, scan response payload.
+
+     * Advertisement payload `allow`
+
+     * Advertisement type `command`
+
+     * If the advertising type is scannable, you SHOULD configure a scan
+       response payload using `allow`
+
+     * Advertisement interval `command`
+
+     * Advertisement tx power `command`
+
+  2. Start periodic advertising using a `command`
+
+Any changes to the configuration while periodic advertising is happening will
+take effect in a future advertising event. The kernel will use best effort to
+reconfigure advertising in as few events as possible.
+
+To stop advertising
+
+  1. Stop periodic advertising a `command`
+
+### Scanning
+
+### Connection-oriented communication
+
+## Hardware Interface Layer (HIL)
+
+The Bluetooth Low Energy Radio HIL defines a cross-platform interface for
+interacting with on-chip BLE radios (i.e. it does not necessarily work for
+radios on a dedicated IC connected over a bus).
+
+The goal of this interface is to expose low-level details of the radio that are
+common across platforms, except in cases where abstraction is needed for common
+cases to meet timing constraints.
+
+
+```rust
+pub trait BleRadio {
+    /// Sets the channel on which to transmit or recieve packets.
+    ///
+    /// Returns ReturnCode::EBUSY if the radio is currently transmiting or
+    /// recieving, otherwise ReturnCode::Success.
+    fn set_channel(&self, channel: RadioChannel) -> ReturnCode;
+
+    /// Sets the transmit power
+    ///
+    /// Returns ReturnCode::EBUSY if the radio is currently transmiting or
+    /// recieving, otherwise ReturnCode::Success.
+    fn set_tx_power(&self, power: u8) -> ReturnCode;
+
+    /// Transmits a packet over the radio
+    ///
+    /// Returns ReturnCode::EBUSY if the radio is currently transmiting or
+    /// recieving, otherwise ReturnCode::Success.
+    fn transmit_packet(
+        &self,
+        buf: &'static mut [u8],
+        disable: bool,
+        len: usize) -> ReturnCode;
```

Comment:
remove `len` I think `buf.len()` is sufficent here!

Reply:
I may be wrong, but I think `len` is required if it's a statically sized buffer (eg. [64; 0] would yield `buf.len() = 64`)

Reply:
No, this is a [slice](https://doc.rust-lang.org/std/primitive.slice.html) not an array

Reply:
Cool, I just assumed that slices were arrays